saleor/discount/sale_converter.py

```diff
@@ -0,0 +1,259 @@
+    )
+
+
+def _create_promotion_rule(
+    sale, promotion, discount_value=None, old_channel_listing_id=None
+):
+    return PromotionRule(
+        name="",
+        promotion=promotion,
+        catalogue_predicate=_create_catalogue_predicate_from_sale(sale),
+        reward_value_type=sale.type,
+        reward_value=discount_value,
+        old_channel_listing_id=old_channel_listing_id,
+    )
+
+
+def _create_catalogue_predicate_from_sale(sale):
+    collection_ids = [
+        graphene.Node.to_global_id("Collection", pk)
+        for pk in sale.collections.values_list("pk", flat=True)
+    ]
+    category_ids = [
+        graphene.Node.to_global_id("Category", pk)
+        for pk in sale.categories.values_list("pk", flat=True)
+    ]
+    product_ids = [
+        graphene.Node.to_global_id("Product", pk)
+        for pk in sale.products.values_list("pk", flat=True)
+    ]
+    variant_ids = [
+        graphene.Node.to_global_id("ProductVariant", pk)
+        for pk in sale.variants.values_list("pk", flat=True)
+    ]
+    return _create_catalogue_predicate(
+        collection_ids, category_ids, product_ids, variant_ids
+    )
+
+
+def _create_catalogue_predicate(collection_ids, category_ids, product_ids, variant_ids):
+    predicate: Dict[str, List] = {"OR": []}
+    if collection_ids:
+        predicate["OR"].append({"collectionPredicate": {"ids": collection_ids}})
+    if category_ids:
+        predicate["OR"].append({"categoryPredicate": {"ids": category_ids}})
+    if product_ids:
+        predicate["OR"].append({"productPredicate": {"ids": product_ids}})
+    if variant_ids:
+        predicate["OR"].append({"variantPredicate": {"ids": variant_ids}})
+
+    return predicate
+
+
+def _migrate_sales_to_promotions(sales_pks, saleid_promotion_map):
+    if sales := Sale.objects.filter(pk__in=sales_pks).order_by("pk"):
+        for sale in sales:
+            saleid_promotion_map[sale.id] = _convert_sale_into_promotion(sale)
+        Promotion.objects.bulk_create(saleid_promotion_map.values())
+
+
+def _migrate_sale_listing_to_promotion_rules(
+    sale_listings,
+    saleid_promotion_map,
+    rules_info,
+):
+    if sale_listings:
+        for sale_listing in sale_listings:
+            promotion = saleid_promotion_map[sale_listing.sale_id]
+            rules_info.append(
+                RuleInfo(
+                    rule=_create_promotion_rule(
+                        sale_listing.sale,
+                        promotion,
+                        sale_listing.discount_value,
+                        sale_listing.id,
+                    ),
+                    sale_id=sale_listing.sale_id,
+                    channel_id=sale_listing.channel_id,
+                )
+            )
+
+        promotion_rules = [rules_info.rule for rules_info in rules_info]
+        PromotionRule.objects.bulk_create(promotion_rules)
+        for rule_info in rules_info:
+            rule_info.add_rule_to_channel()
```

Comment:
Are you sure about using add_rule_to_channel() here? It looks like it will create a single query for each `rule_info`.
Shouldn't we call bulk_update here?  
I think you could do something like:
`PromotionRuleChannel = PromotionRule.channels.through`

Reply:
I think, the same thing is in 0049 migration

Reply:
Thanks, I was not aware there is such an option